sdk/python/approzium/psycopg2.py

```diff
@@ -45,107 +85,128 @@ def intataddress(address):
+        elif state == psycopg2.extensions.POLL_READ:
+            select.select([pgconn.fileno()], [], [])
+        else:
+            raise psycopg2.OperationalError("poll() returned %s" % state)
 
 
 def send_hash(pgconn, hash):
-    sock = fromfd(pgconn.fileno(), keep_fd=True)
     msg_length = 40  # fixed number for MD5 hash result
     # XXX: following code assumes protocol 3
     msg = b"p"  # message type
     msg += struct.pack("!i", msg_length)
     msg += b"md5"
     msg += hash.encode("ascii")
     msg += b"\0"
-    if libpq.PQsslInUse(pgconn.pgconn_ptr):
-        ssl_obj = libpq.PQgetssl(pgconn.pgconn_ptr)
+    if libpq_PQsslInUse(pgconn.pgconn_ptr):
+        ssl_obj = libpq_PQgetssl(pgconn.pgconn_ptr)
         c_buffer = create_string_buffer(msg, len(msg))
-        n = libssl.SSL_write(ssl_obj, c_buffer, len(msg))
+        n = libssl_SSL_write(ssl_obj, c_buffer, len(msg))
         if n != len(msg):
             raise ValueError("could not send response")
     else:
-        sock.sendall(msg)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", ResourceWarning)
+            sock = fromfd(pgconn.fileno(), keep_fd=True)
+            sock.sendall(msg)
+
+
+def construct_approzium_conn(base, is_sync):
+    if not base:
+        base = psycopg2.extensions.connection
+
+    class ApproziumConn(base):
+        CONNECTION_AWAITING_RESPONSE = 4
+
+        def __init__(self, *args, **kwargs):
+            # can safely do so because real async value was caught earlier in our connect method
+            logging.debug("ApproziumConn __init__")
+            kwargs.pop("async", None)
+            kwargs.pop("async_", None)
+            conn = super().__init__(*args, **kwargs, async_=1)
+            if self.dsn is None:
+                # connection is uninitalized due to an error
+                return
+            self._salt = None
+            self._hash_sent = False
+            if is_sync:
+                wait(self)
+                set_connection_sync(self)
+                self.autocommit = False
+
+        def poll(self):
+            status = libpq_PQstatus(self.pgconn_ptr)
+            if status == self.CONNECTION_AWAITING_RESPONSE and not self._salt:
+                logging.debug("reading salt")
+                self._salt = read_salt(self)
+                return psycopg2.extensions.POLL_WRITE
+            elif self._salt and not self._hash_sent:
+                logging.debug("sending hash")
+                dbhost = self.get_dsn_parameters()["host"]
+                dbuser = self.get_dsn_parameters()["user"]
+                hash = get_hash(
+                    dbhost, dbuser, self._salt, approzium.authenticator_addr
+                )
+                send_hash(self, hash)
+                self._hash_sent = True
+                return psycopg2.extensions.POLL_WRITE
+            else:
+                logging.debug("normal poll")
+                return super().poll()
```

Comment:
overriding the `poll` method allows us to allow the user to create the connection in an async manner.